### DISCUSSION TITLE
Inifite loop with watch

### DISCUSSION THREAD

**Discussion by pedroteixeira** [pmndrs/valtio#720] [General]
Hello! I wanted to bring this to attention to get some ideas.
I was migrating from 1.3.1, and some previous watch implementations now get suck in infinite loop (it happens as early as in 1.4.0)

Noticed the issue happens because implementation was creating new `{}` and assigning to a key inside `watch`:

```
    it('simulates infinite loop in watching', async () => {
        const store = proxy({ options: {} })

        watch(get => {
            const state = get(store)

            store.derivedKey = state.options[state.key] || {}
        })

        store.key = 'x'

        await Promise.resolve()
    })
```

I am not sure best way the library can help in such cases, just bringing here for discussion :)

Perhaps a development version that uses "deep equals" to warn that an equivalent but different object is being re-assigned multiple times?

Thanks

**Comment by dai-shi** [comment]
As a workaround, does defining `const EMPTY = {};` outside solves the issue?

> (it happens as early as in 1.4.0)

So, #385?

I wonder how 1.3.1 worked... 🤔

**Reply by pedroteixeira** [reply]
Yes, that's what I ended up doing to avoid the issue (using a constant empty object when initializing in similar scenario, but may not always be possible if not readonly, so just extra care to create a new object once if undefined, and then perhaps mutate other properties after object creation)

Just thought about bringing it here for awareness - we might want to add some "in development" checks
since developers need to be extra careful when using watch, due to this possible behavior

**Comment by dai-shi** [comment]
If anyone is interested, feel free to work on it: fix it potentially or otherwise possibly warn it in DEV.

**Comment by gladimir777** [comment]
Hello guys, is this issue still open. If not. I would like to work on it. And start contributing on open source project. Thank you!

**Reply by dai-shi** [reply]
It's open, but I'm not sure if it's easily fixable or if even we should fix it.
If it's your first time contribution, you might want to look for a simpler issue.

**Reply by gladimir777** [reply]
Thank you for the advice. and really inspired by your work on zustand. I would like to be able to understand and contribute to open source like this and zustand. I will look at other projects and find some easy ones

**Reply by overthemike** [reply]
> It's open, but I'm not sure if it's easily fixable or if even we should fix it. If it's your first time contribution, you might want to look for a simpler issue.

I don't think the watch function should be "fixed" to do this. If anything, there could be a separate utility lib called something like "watchByValue" to do this. Maybe we could create another discussion topic for "feature/library requests" for stuff like this to be written down where someone might take it on?

**Reply by dai-shi** [reply]
alright, let's close this then.
